src/engine/controls/macrocontrol.cpp

```diff
@@ -0,0 +1,262 @@
+#include "macrocontrol.h"
+
+namespace {
+constexpr uint kRecordingTimerInterval = 100;
+constexpr size_t kRecordingQueueSize = kRecordingTimerInterval / 10;
+} // namespace
+
+ConfigKey MacroControl::getConfigKey(QString name) {
+    return ConfigKey(m_group, m_controlPattern.arg(name));
+}
+
+MacroControl::MacroControl(QString group, UserSettingsPointer pConfig, int slot)
+        : EngineControl(group, pConfig),
+          m_slot(slot),
+          m_controlPattern(QString("macro_%1_%2").arg(slot)),
+          m_bJumpPending(false),
+          m_recordedActions(kRecordingQueueSize),
+          m_iNextAction(0),
+          m_COPlaying(getConfigKey("playing")),
+          m_COStatus(getConfigKey("status")),
+          m_record(getConfigKey("record")),
+          m_toggle(getConfigKey("toggle")),
```

Comment:
This interface still does not fit to Mixxx. All controls have a build in behavior, that should be used to be easily map-able to GUI buttons and midi controllers. 

I think this can become a "play" control with toggle button mode, a joined value with "playing"

Reply:
The problem here is that `toggle` can also record, so in a sense it is an alternative mapping for `activate`.